griptape/tools/email_client/tool.py

```diff
@@ -139,51 +140,83 @@ def retrieve(self, params: dict) -> ListArtifact | ErrorArtifact:
             ): str,
             Literal(
                 "body",
-                description="Email body"
-            ): str
+                description="Email body")
+            : str,
+            Optional(
+                "attachment_names",
+                description="Names of the attachments"
+            ): list[str],
+            Optional(
+                "memory_name",
+                description="Names of the attachments"
+            ): str,
+            Optional(
+                "artifact_namespace",
+                description="Names of the attachments"
+            ): str,
         })
     })
     def send(self, params: dict) -> InfoArtifact | ErrorArtifact:
         input_values = params["values"]
         server: Optional[smtplib.SMTP] = None
+        to_email = input_values.get("to")
+        subject = input_values.get("subject")
+        memory_name = input_values.get("memory_name")
+        artifact_namespace = input_values.get("artifact_namespace")
+        attachment_names = input_values.get("attachment_names")
 
         # email username can be overridden by setting the smtp user explicitly
-        smtp_user = self.smtp_user
-        if smtp_user is None:
-            smtp_user = self.username
+        smtp_user = self.smtp_user if self.smtp_user else self.username
 
         # email password can be overridden by setting the smtp password explicitly
-        smtp_password = self.smtp_password
-        if smtp_password is None:
-            smtp_password = self.password
+        smtp_password = self.smtp_password if self.smtp_password else self.password
 
         smtp_host = self.smtp_host
         smtp_port = int(self.smtp_port)
 
-        to_email = input_values["to"]
-        subject = input_values["subject"]
-        msg = MIMEText(input_values["body"])
+        # Create a multipart email and set headers
+        msg = MIMEMultipart()
+        msg["From"] = smtp_user
+        msg["To"] = to_email
+        msg["Subject"] = subject
+
+        # Attach the body to the email
+        msg.attach(MIMEText(input_values["body"]))
+
+        # Fetch attachment data from memory
+        if memory_name and artifact_namespace:
+            memory = self.find_input_memory(memory_name)
+            if memory:
+                list_artifact = memory.load_artifacts(artifact_namespace)
+                if list_artifact:
+                    for artifact, attachment_name in zip(list_artifact.value, attachment_names):
+                        file_data = artifact.value.encode()
+                        part = MIMEBase("application", "octet-stream")
+                        part.set_payload(file_data)
+                        encoders.encode_base64(part)
+                        part.add_header("Content-Disposition", f"attachment; filename={attachment_name}")
+                        msg.attach(part)
+                else:
+                    return ErrorArtifact(f"Artifact with namespace {artifact_namespace} not found.")
```

Comment:
Nesting gets pretty deep here, can we refactor this? Perhaps split out into some separate functions?

Reply:
made a few protected methods